Replace +[WKPreferences _experimentalFeatures] and +[WKPreferences _internalDebugFeatures] with SPI for exposing all features

emw-apple opened the pull request:
#### 81a368e50d6eb22af62d70a23ffa2b7d7ad3d7cc
<pre>
Replace +[WKPreferences _experimentalFeatures] and +[WKPreferences _internalDebugFeatures] with SPI for exposing all features
<a href="https://bugs.webkit.org/show_bug.cgi?id=250237">https://bugs.webkit.org/show_bug.cgi?id=250237</a>
rdar://103648153

Reviewed by Brent Fulgham.

Add +[WKPreferences _features] and WebPreferences::features(), which
exposes the entire feature list, now that features are unified and distinguished by
status. Prepare to deprecate other ways to access the feature list from
WebKit.

Unify the internal and experimental feature classes into one class.
API::ExperimentalFeature and API::InternalDebugFeature are now
&quot;API::Feature&quot;. In Cocoa SPI, _WKExperimentalFeature and
_WKInternalDebugFeature are now &quot;_WKFeature&quot;, though the old classes
have been left around for compatibility.

Rename-only file changes omitted from this change list:

* Source/WTF/Scripts/GeneratePreferences.rb:
* Source/WebKit/Scripts/PreferencesTemplates/WebPreferencesFeatures.cpp.erb:
  Renamed from Source/WebKit/Scripts/PreferencesTemplates/WebPreferencesExperimentalFeatures.cpp.erb.
  Instead of generating separate &quot;experimental&quot; and &quot;internal debug&quot;
  vectors, it now generates one vector for WebPreferences::features().
* Source/WebKit/Scripts/PreferencesTemplates/WebPreferencesInternalDebugFeatures.cpp.erb: Removed.
* Source/WebKit/UIProcess/API/APIExperimentalFeature.h: Removed.
* Source/WebKit/UIProcess/API/APIFeature.cpp: Renamed from Source/WebKit/UIProcess/API/APIExperimentalFeature.cpp.
* Source/WebKit/UIProcess/API/APIFeature.h: Renamed from Source/WebKit/UIProcess/API/APIInternalDebugFeature.h.
* Source/WebKit/UIProcess/API/APIInternalDebugFeature.cpp: Removed.
* Source/WebKit/UIProcess/API/C/WKFeature.cpp: Renamed from Source/WebKit/UIProcess/API/C/WKExperimentalFeature.cpp.
* Source/WebKit/UIProcess/API/C/WKFeature.h: Renamed from Source/WebKit/UIProcess/API/C/WKExperimentalFeature.h.
* Source/WebKit/UIProcess/API/C/WKInternalDebugFeature.cpp: Removed.
* Source/WebKit/UIProcess/API/C/WKInternalDebugFeature.h: Removed.
* Source/WebKit/UIProcess/API/C/WKPreferences.cpp:
* Source/WebKit/UIProcess/API/Cocoa/WKPreferences.mm:
(+[WKPreferences _features]): Added.

  These methods used to be legacy ways to enable experimental features
  (before internal debug features were added). Repurpose them to work for
  *any* feature.
(-[WKPreferences _isEnabledForFeature:]):
(-[WKPreferences _setEnabled:forFeature:]):

  These are the legacy ways to access features. They should be deprecated
  in a future change, and internally they are refactored to use the above.
(+[WKPreferences _internalDebugFeatures]):
(-[WKPreferences _isEnabledForInternalDebugFeature:]):
(-[WKPreferences _setEnabled:forInternalDebugFeature:]):
(+[WKPreferences _experimentalFeatures]):
(-[WKPreferences _isEnabledForExperimentalFeature:]):
(-[WKPreferences _setEnabled:forExperimentalFeature:]):

* Source/WebKit/UIProcess/API/Cocoa/WKPreferencesPrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKExperimentalFeature.h: Stub
  subclass of _WKFeature for binary compatibility.
* Source/WebKit/UIProcess/API/Cocoa/_WKExperimentalFeature.mm:
(-[_WKExperimentalFeature dealloc]): Deleted.
(-[_WKExperimentalFeature description]): Deleted.
(-[_WKExperimentalFeature name]): Deleted.
(-[_WKExperimentalFeature status]): Deleted.
(-[_WKExperimentalFeature key]): Deleted.
(-[_WKExperimentalFeature details]): Deleted.
(-[_WKExperimentalFeature defaultValue]): Deleted.
(-[_WKExperimentalFeature isHidden]): Deleted.
(-[_WKExperimentalFeature _apiObject]): Deleted.
* Source/WebKit/UIProcess/API/Cocoa/_WKFeature.h: Added.
* Source/WebKit/UIProcess/API/Cocoa/_WKFeature.mm: Copied from Source/WebKit/UIProcess/API/Cocoa/_WKExperimentalFeature.mm.
(-[_WKFeature _apiObject]):
* Source/WebKit/UIProcess/API/Cocoa/_WKFeatureInternal.h: Renamed from Source/WebKit/UIProcess/API/Cocoa/_WKExperimentalFeatureInternal.h.
* Source/WebKit/UIProcess/API/Cocoa/_WKInternalDebugFeature.h: Stub
  subclass of _WKFeature for binary compatibility.
* Source/WebKit/UIProcess/API/Cocoa/_WKInternalDebugFeature.mm:
(-[_WKInternalDebugFeature dealloc]): Deleted.
(-[_WKInternalDebugFeature description]): Deleted.
(-[_WKInternalDebugFeature name]): Deleted.
(-[_WKInternalDebugFeature status]): Deleted.
(-[_WKInternalDebugFeature key]): Deleted.
(-[_WKInternalDebugFeature details]): Deleted.
(-[_WKInternalDebugFeature defaultValue]): Deleted.
(-[_WKInternalDebugFeature isHidden]): Deleted.
(-[_WKInternalDebugFeature _apiObject]): Deleted.
* Source/WebKit/UIProcess/API/Cocoa/_WKInternalDebugFeatureInternal.h: Removed.

* Source/WebKit/UIProcess/WebPreferences.cpp:
(WebKit::WebPreferences::experimentalFeatures): This vector used to be
  produced by the source generator. Now, it&apos;s initialized once at runtime
  by filtering the list of features.
(WebKit::WebPreferences::internalDebugFeatures): Ditto.
* Source/WebKit/UIProcess/WebPreferences.h:

* Tools/MiniBrowser/mac/AppDelegate.m: Updated to use
  +[WKPreferences _features] and the corresponding setter/getter.
* Tools/MiniBrowser/mac/SettingsController.m: Ditto.
* Tools/MiniBrowser/win/WebKitBrowserWindow.cpp: Use updated C API.

TestWebKitAPI is largely changed over to use +[WKPreferences _features]
instead of one of the legacy interfaces. Other changes include:

* Tools/TestWebKitAPI/Tests/WebKitCocoa/ProcessSwapOnNavigation.mm:
  This test used to turn on all &quot;experimental&quot; features to test whether
  some were disabled in Lockdown mode. Now, it turns on all testable and
  preview features, as those are closer to what LDM actually disables.

Canonical link: <a href="https://commits.webkit.org/258746@main">https://commits.webkit.org/258746@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/697715b2274c3d90359a1655b66538835a1a29a9

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/102849 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/77/builds/11967 "Build was cancelled. Recent messages:Cleaned up git repository; Skipping applying patch since patch_id isn't provided; Checked out pull request; Compiled WebKit (cancelled)") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/35875 "Build was cancelled. Recent messages:Cleaned up git repository; Skipping applying patch since patch_id isn't provided; Checked out pull request; Compiled WebKit (cancelled)") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/112103 "Built successfully") | [💥 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/172323 "An unexpected error occured. Recent messages:Pull request contains relevant changes; Failed to print configuration") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/12984 "Built successfully") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/85/builds/2876 "Build was cancelled. Recent messages:Pull request contains relevant changes; Cleaned up git repository; Skipping applying patch since patch_id isn't provided; Checked out pull request; Compiled WebKit (warnings); Failed to upload built product") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/95097 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/109773 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/108623 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/9976 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/35875 "Build was cancelled. Recent messages:Cleaned up git repository; Skipping applying patch since patch_id isn't provided; Checked out pull request; Compiled WebKit (cancelled)") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/37613 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/5/builds/106362 "Passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/91821 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/35875 "Build was cancelled. Recent messages:Cleaned up git repository; Skipping applying patch since patch_id isn't provided; Checked out pull request; Compiled WebKit (cancelled)") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/79347 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/1/builds/93055 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/5420 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/35875 "Build was cancelled. Recent messages:Cleaned up git repository; Skipping applying patch since patch_id isn't provided; Checked out pull request; Compiled WebKit (cancelled)") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/89417 "Built successfully") | 
| [  ~~🛠 🧪 jsc-arm64~~](https://ews-build.webkit.org/#/builders/83/builds/3106 "Build was cancelled. Recent messages:Pull request contains relevant changes; Cleaned up git repository; Skipping applying patch since patch_id isn't provided; Checked out pull request; Compiled JSC (warnings); jscore-tests (cancelled)") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/5568 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/85/builds/2876 "Build was cancelled. Recent messages:Pull request contains relevant changes; Cleaned up git repository; Skipping applying patch since patch_id isn't provided; Checked out pull request; Compiled WebKit (warnings); Failed to upload built product") | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/29677 "Found 1 new JSC stress test failure: wasm.yaml/wasm/lowExecutableMemory/exports-oom.js.default-wasm (failure)") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/80/builds/11587 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/35875 "Build was cancelled. Recent messages:Cleaned up git repository; Skipping applying patch since patch_id isn't provided; Checked out pull request; Compiled WebKit (cancelled)") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/92342 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6011 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/7313 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/20680 "Passed tests") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/WebKit/DerivedSources.make:
```diff
@@ -441,9 +441,8 @@ WEB_PREFERENCES = \
     $(WebKit2)/Scripts/PreferencesTemplates/WebPageUpdatePreferences.cpp.erb \
     $(WebKit2)/Scripts/PreferencesTemplates/WebPreferencesDefinitions.h.erb \
-    $(WebKit2)/Scripts/PreferencesTemplates/WebPreferencesExperimentalFeatures.cpp.erb \
+    $(WebKit2)/Scripts/PreferencesTemplates/WebPreferencesFeatures.cpp.erb \
```
You need to make a similar change to CMake’s file.

brentfulgham commented on Source/WebKit/Scripts/PreferencesTemplates/WebPreferencesFeatures.cpp.erb:
```diff
@@ -126,4 +126,19 @@ void WebPreferences::enableAllExperimentalFeatures()
+<%- end -%>
+}
+
+
```
Nit: One extra line of whitespace, here.

brentfulgham commented on Source/WebKit/UIProcess/API/APIFeature.h:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (C) 2018 Apple Inc. All rights reserved.
+ * Copyright (C) 2016 Apple Inc. All rights reserved.
```
Instead of `(C) 2016`, it should be `(C) 2016-2023`, instead.

brentfulgham commented on Source/WebKit/UIProcess/WebPreferences.cpp:
```diff
@@ -77,6 +77,34 @@ WebPreferences::~WebPreferences()
+        Vector<RefPtr<API::Object>> result;
+        for (auto& object : features()) {
+            API::FeatureStatus status = static_pointer_cast<API::Feature>(object)->status();
+            if (status == API::FeatureStatus::Unstable || status == API::FeatureStatus::Testable)
```
I think this should be:

`API::FeatureStatus::Unstable || API::FeatureStatus::Internal`

emw-apple replied:
That's right, good catch!

brentfulgham reviewed (approved):
Looks good. Do you need to coordinate with Patrick for this change?

Also: I think you need to correct the Internal Debug case (as mentioned below).

emw-apple committed:
Replace +[WKPreferences _experimentalFeatures] and +[WKPreferences _internalDebugFeatures] with SPI for exposing all features
https://bugs.webkit.org/show_bug.cgi?id=250237
rdar://103648153

Reviewed by Brent Fulgham.

Add +[WKPreferences _features] and WebPreferences::features(), which
exposes the entire feature list, now that features are unified and distinguished by
status. Prepare to deprecate other ways to access the feature list from
WebKit.

Unify the internal and experimental feature classes into one class.
API::ExperimentalFeature and API::InternalDebugFeature are now
"API::Feature". In Cocoa SPI, _WKExperimentalFeature and
_WKInternalDebugFeature are now "_WKFeature", though the old classes
have been left around for compatibility.

Rename-only file changes omitted from this change list:

* Source/WTF/Scripts/GeneratePreferences.rb:
* Source/WebKit/Scripts/PreferencesTemplates/WebPreferencesFeatures.cpp.erb:
  Renamed from Source/WebKit/Scripts/PreferencesTemplates/WebPreferencesExperimentalFeatures.cpp.erb.
  Instead of generating separate "experimental" and "internal debug"
  vectors, it now generates one vector for WebPreferences::features().
* Source/WebKit/Scripts/PreferencesTemplates/WebPreferencesInternalDebugFeatures.cpp.erb: Removed.
* Source/WebKit/UIProcess/API/APIExperimentalFeature.h: Removed.
* Source/WebKit/UIProcess/API/APIFeature.cpp: Renamed from Source/WebKit/UIProcess/API/APIExperimentalFeature.cpp.
* Source/WebKit/UIProcess/API/APIFeature.h: Renamed from Source/WebKit/UIProcess/API/APIInternalDebugFeature.h.
* Source/WebKit/UIProcess/API/APIInternalDebugFeature.cpp: Removed.
* Source/WebKit/UIProcess/API/C/WKFeature.cpp: Renamed from Source/WebKit/UIProcess/API/C/WKExperimentalFeature.cpp.
* Source/WebKit/UIProcess/API/C/WKFeature.h: Renamed from Source/WebKit/UIProcess/API/C/WKExperimentalFeature.h.
* Source/WebKit/UIProcess/API/C/WKInternalDebugFeature.cpp: Removed.
* Source/WebKit/UIProcess/API/C/WKInternalDebugFeature.h: Removed.
* Source/WebKit/UIProcess/API/C/WKPreferences.cpp:
* Source/WebKit/UIProcess/API/Cocoa/WKPreferences.mm:
(+[WKPreferences _features]): Added.

  These methods used to be legacy ways to enable experimental features
  (before internal debug features were added). Repurpose them to work for
  *any* feature.
(-[WKPreferences _isEnabledForFeature:]):
(-[WKPreferences _setEnabled:forFeature:]):

  These are the legacy ways to access features. They should be deprecated
  in a future change, and internally they are refactored to use the above.
(+[WKPreferences _internalDebugFeatures]):
(-[WKPreferences _isEnabledForInternalDebugFeature:]):
(-[WKPreferences _setEnabled:forInternalDebugFeature:]):
(+[WKPreferences _experimentalFeatures]):
(-[WKPreferences _isEnabledForExperimentalFeature:]):
(-[WKPreferences _setEnabled:forExperimentalFeature:]):

* Source/WebKit/UIProcess/API/Cocoa/WKPreferencesPrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKExperimentalFeature.h: Stub
  subclass of _WKFeature for binary compatibility.
* Source/WebKit/UIProcess/API/Cocoa/_WKExperimentalFeature.mm:
(-[_WKExperimentalFeature dealloc]): Deleted.
(-[_WKExperimentalFeature description]): Deleted.
(-[_WKExperimentalFeature name]): Deleted.
(-[_WKExperimentalFeature status]): Deleted.
(-[_WKExperimentalFeature key]): Deleted.
(-[_WKExperimentalFeature details]): Deleted.
(-[_WKExperimentalFeature defaultValue]): Deleted.
(-[_WKExperimentalFeature isHidden]): Deleted.
(-[_WKExperimentalFeature _apiObject]): Deleted.
* Source/WebKit/UIProcess/API/Cocoa/_WKFeature.h: Added.
* Source/WebKit/UIProcess/API/Cocoa/_WKFeature.mm: Copied from Source/WebKit/UIProcess/API/Cocoa/_WKExperimentalFeature.mm.
(-[_WKFeature _apiObject]):
* Source/WebKit/UIProcess/API/Cocoa/_WKFeatureInternal.h: Renamed from Source/WebKit/UIProcess/API/Cocoa/_WKExperimentalFeatureInternal.h.
* Source/WebKit/UIProcess/API/Cocoa/_WKInternalDebugFeature.h: Stub
  subclass of _WKFeature for binary compatibility.
* Source/WebKit/UIProcess/API/Cocoa/_WKInternalDebugFeature.mm:
(-[_WKInternalDebugFeature dealloc]): Deleted.
(-[_WKInternalDebugFeature description]): Deleted.
(-[_WKInternalDebugFeature name]): Deleted.
(-[_WKInternalDebugFeature status]): Deleted.
(-[_WKInternalDebugFeature key]): Deleted.
(-[_WKInternalDebugFeature details]): Deleted.
(-[_WKInternalDebugFeature defaultValue]): Deleted.
(-[_WKInternalDebugFeature isHidden]): Deleted.
(-[_WKInternalDebugFeature _apiObject]): Deleted.
* Source/WebKit/UIProcess/API/Cocoa/_WKInternalDebugFeatureInternal.h: Removed.

* Source/WebKit/UIProcess/WebPreferences.cpp:
(WebKit::WebPreferences::experimentalFeatures): This vector used to be
  produced by the source generator. Now, it's initialized once at runtime
  by filtering the list of features.
(WebKit::WebPreferences::internalDebugFeatures): Ditto.
* Source/WebKit/UIProcess/WebPreferences.h:

* Tools/MiniBrowser/mac/AppDelegate.m: Updated to use
  +[WKPreferences _features] and the corresponding setter/getter.
* Tools/MiniBrowser/mac/SettingsController.m: Ditto.
* Tools/MiniBrowser/win/WebKitBrowserWindow.cpp: Use updated C API.

TestWebKitAPI is largely changed over to use +[WKPreferences _features]
instead of one of the legacy interfaces. Other changes include:

* Tools/TestWebKitAPI/Tests/WebKitCocoa/ProcessSwapOnNavigation.mm:
  This test used to turn on all "experimental" features to test whether
  some were disabled in Lockdown mode. Now, it turns on all testable and
  preview features, as those are closer to what LDM actually disables.

Canonical link: https://commits.webkit.org/258746@main
